File: Chapter02_LinkedList/02-06_kaibun_list/02-06_kaibun_list.py
# -*- coding: utf-8 -*-
class ListNode:
     def __init__(self, value="", next=None):
          self.value = value
          self.next = next

def main():
     # [1]->[2]->[2]->[3]->[1]->[4]
     len_node = 5
     n5 = ListNode("🐟")
     n4 = ListNode("💖", n5)
     n3 = ListNode("🌻", n4)
     n2 = ListNode("💖", n3)
     n1 = ListNode("🐟", n2)
     head = n1
     stack = []
     count = 0
     while count < (int)(len_node/2):
          stack.append(head.value)
          head = head.next
          count += 1
     if len_node % 2 == 1:
          head = head.next

     while head != None:
          if head.value != stack.pop():
               print("False")
               return
          head = head.next
     print("True")
if __name__ == "__main__":
     main()

File: Chapter02_LinkedList/02-06_kaibun_list/02-06_kaibun_list.cpp
#include<iostream>
#include<stack>

using namespace std;

class Node {
public:
    char val;
    Node* next;
    Node(int v) {
      val = v;
    }
};

bool isKaibun(Node* head, int length) {
    stack<char> st;
    Node* curr = head;

    for (int i = 0; i < length/2; i++)   {
      st.push(curr->val);
      curr = curr->next;
  }

  if (length % 2) curr = curr->next;

  /*for (int i = 0; i < length/2; i++) {
    char top = st.top();
    st.pop();
    if (top != curr->val) return false;
  }*/
  return true;
}

int main () {
    string s = "madam";
    Node* head = new Node('-');
    Node* curr = head;
    for (int i = 0; i < s.size(); i++) {
      Node* tmp = new Node(s[i]);
      curr->next = tmp;
      curr = curr->next;
  }
  if (isKaibun(head->next, s.size())) {
    cout << "true" << endl;
  } else {
    cout << "false" << endl;
  }
}